File: ios/Antislot/SharedConfigModule.swift
import Foundation
import React

@objc(SharedConfigModule)
class SharedConfigModule: NSObject, RCTBridgeModule {
  static func moduleName() -> String! {
    return "SharedConfigModule"
  }

  static func requiresMainQueueSetup() -> Bool {
    return false
  }

  private let appGroupId = "group.com.antislot.app"
  private let keySmsSettings = "sms.settings"
  private let keySmsEnabled = "sms.enabled"
  private let keySmsStrictMode = "sms.strictMode"
  private let keySmsKeywords = "sms.keywords"
  private let keySmsAutoDeleteDays = "sms.autoDeleteDays"
  private let keySmsBlockedCount = "sms.blockedCount"

  private func storage() -> UserDefaults {
    return UserDefaults(suiteName: appGroupId) ?? UserDefaults.standard
  }

  @objc func saveBlocklist(
    _ domains: [String],
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    storage().set(domains, forKey: "blocker.blocklist")
    resolve(true)
  }

  @objc func savePatterns(
    _ patterns: NSArray,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    storage().set(patterns, forKey: "blocker.patterns")
    resolve(true)
  }

  @objc func saveWhitelist(
    _ domains: [String],
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    storage().set(domains, forKey: "blocker.whitelist")
    resolve(true)
  }

  @objc func saveBlockerHardening(
    _ strictMode: Bool,
    blockDoh: Bool,
    blockDot: Bool,
    blockQuic: Bool,
    lockdownVpn: Bool,
    tamperAlerts: Bool,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    let settings: [String: Any] = [
      "strictMode": strictMode,
      "blockDoh": blockDoh,
      "blockDot": blockDot,
      "blockQuic": blockQuic,
      "lockdownVpn": lockdownVpn,
      "tamperAlerts": tamperAlerts
    ]
    storage().set(settings, forKey: "blocker.hardening")
    resolve(true)
  }

  @objc func saveSmsSettings(
    _ enabled: Bool,
    strictMode: Bool,
    customKeywords: [String],
    autoDeleteDays: NSNumber,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    let normalizedKeywords = customKeywords
      .map { $0.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() }
      .filter { !$0.isEmpty }

    let prefs: [String: Any] = [
      "enabled": enabled,
      "strictMode": strictMode,
      "customKeywords": normalizedKeywords,
      "autoDeleteDays": autoDeleteDays
    ]
    let defaults = storage()
    defaults.set(prefs, forKey: keySmsSettings)
    defaults.set(enabled, forKey: keySmsEnabled)
    defaults.set(strictMode, forKey: keySmsStrictMode)
    defaults.set(normalizedKeywords, forKey: keySmsKeywords)
    if autoDeleteDays.intValue >= 0 {
      defaults.set(autoDeleteDays.intValue, forKey: keySmsAutoDeleteDays)
    } else {
      defaults.removeObject(forKey: keySmsAutoDeleteDays)
    }
    defaults.synchronize()
    resolve(true)
  }

  @objc func getSmsBlockedCount(
    _ resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    let blocked = storage().integer(forKey: keySmsBlockedCount)
    resolve(blocked)
  }

  @objc func resetSmsBlockedCount(
    _ resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    let defaults = storage()
    defaults.removeObject(forKey: keySmsBlockedCount)
    defaults.synchronize()
    resolve(true)
  }

  @objc func cleanupSpamInboxNow(
    _ resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    resolve(0)
  }
}

File: android/app/src/main/java/com/antislot/SharedConfigModule.kt
package com.antislot

import android.util.Log
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.ReadableType
import org.json.JSONArray

class SharedConfigModule(reactContext: ReactApplicationContext) :
  ReactContextBaseJavaModule(reactContext) {

  private val tag = "SharedConfigModule"

  override fun getName(): String = "SharedConfigModule"

  @ReactMethod
  fun saveBlocklist(domains: ReadableArray, promise: Promise) {
    try {
      SharedConfigStore.saveBlocklist(reactApplicationContext, toStringList(domains))
      promise.resolve(true)
    } catch (error: Exception) {
      Log.e(tag, "saveBlocklist failed", error)
      promise.resolve(false)
    }
  }

  @ReactMethod
  fun savePatterns(patterns: ReadableArray, promise: Promise) {
    try {
      SharedConfigStore.savePatterns(
        reactApplicationContext,
        JSONArray(patterns.toArrayList()).toString()
      )
      promise.resolve(true)
    } catch (error: Exception) {
      Log.e(tag, "savePatterns failed", error)
      promise.resolve(false)
    }
  }

  @ReactMethod
  fun saveWhitelist(domains: ReadableArray, promise: Promise) {
    try {
      SharedConfigStore.saveWhitelist(reactApplicationContext, toStringList(domains))
      promise.resolve(true)
    } catch (error: Exception) {
      Log.e(tag, "saveWhitelist failed", error)
      promise.resolve(false)
    }
  }

  @ReactMethod
  fun saveBlockerHardening(
    strictMode: Boolean,
    blockDoh: Boolean,
    blockDot: Boolean,
    blockQuic: Boolean,
    lockdownVpn: Boolean,
    tamperAlerts: Boolean,
    promise: Promise
  ) {
    try {
      SharedConfigStore.saveBlockerHardening(
        context = reactApplicationContext,
        strictMode = strictMode,
        blockDoh = blockDoh,
        blockDot = blockDot,
        blockQuic = blockQuic,
        lockdownVpn = lockdownVpn,
        tamperAlerts = tamperAlerts
      )
      promise.resolve(true)
    } catch (error: Exception) {
      Log.e(tag, "saveBlockerHardening failed", error)
      promise.resolve(false)
    }
  }

  @ReactMethod
  fun saveSmsSettings(
    enabled: Boolean,
    strictMode: Boolean,
    customKeywords: ReadableArray,
    autoDeleteDays: Int,
    promise: Promise
  ) {
    try {
      SharedConfigStore.saveSmsSettings(
        context = reactApplicationContext,
        enabled = enabled,
        strictMode = strictMode,
        keywords = toStringList(customKeywords),
        autoDeleteDays = autoDeleteDays.takeIf { it >= 0 }
      )
      promise.resolve(true)
    } catch (error: Exception) {
      Log.e(tag, "saveSmsSettings failed", error)
      promise.resolve(false)
    }
  }

  @ReactMethod
  fun getSmsBlockedCount(promise: Promise) {
    try {
      promise.resolve(SharedConfigStore.getBlockedCount(reactApplicationContext))
    } catch (error: Exception) {
      Log.e(tag, "getSmsBlockedCount failed", error)
      promise.resolve(0)
    }
  }

  @ReactMethod
  fun resetSmsBlockedCount(promise: Promise) {
    try {
      SharedConfigStore.resetBlockedCount(reactApplicationContext)
      promise.resolve(true)
    } catch (error: Exception) {
      Log.e(tag, "resetSmsBlockedCount failed", error)
      promise.resolve(false)
    }
  }

  @ReactMethod
  fun cleanupSpamInboxNow(promise: Promise) {
    try {
      val deleted = SharedConfigStore.cleanupSpamInboxNow(reactApplicationContext)
      promise.resolve(deleted)
    } catch (error: Exception) {
      Log.e(tag, "cleanupSpamInboxNow failed", error)
      promise.resolve(0)
    }
  }

  private fun toStringList(array: ReadableArray): List<String> {
    val items = mutableListOf<String>()
    for (index in 0 until array.size()) {
      if (array.getType(index) == ReadableType.String) {
        array.getString(index)?.let { value ->
          if (value.isNotBlank()) items += value
        }
      }
    }
    return items
  }
}
